src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c

```diff
@@ -0,0 +1,641 @@
+{
+    JNIEnv* env = GetJNIEnv();
+    jbyteArray buffArray = (*env)->NewByteArray(env, len);
+    (*env)->SetByteArrayRegion(env, buffArray, 0, len, (jbyte*)data);
+    jobject bigNum;
+    if (data[0] & 0x80)
+    {
+        // If the top bit is set, we need to create a BigInteger with the sign-and-magnitude constructor.
+        bigNum = (*env)->NewObject(env, g_bigNumClass, g_bigNumCtorWithSign, 1, buffArray);
+    }
+    else
+    {
+        // Otherwise, we need to create it with the byte-array-value constructor.
+        bigNum = (*env)->NewObject(env, g_bigNumClass, g_bigNumCtor, buffArray);
+    }
+    (*env)->DeleteLocalRef(env, buffArray);
+    return CheckJNIExceptions(env) ? FAIL : ToGRef(env, bigNum);
+}
+
+int32_t CryptoNative_GetECKeyParameters(const EC_KEY* key,
+                                        int32_t includePrivate,
+                                        jobject* qx,
+                                        int32_t* cbQx,
+                                        jobject* qy,
+                                        int32_t* cbQy,
+                                        jobject* d,
+                                        int32_t* cbD)
+{
+    assert(qx != NULL);
+    assert(cbQx != NULL);
+    assert(qy != NULL);
+    assert(cbQy != NULL);
+    assert(d != NULL);
+    assert(cbD != NULL);
+
+    JNIEnv* env = GetJNIEnv();
+
+    // Get the public key
+    jobject publicKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPublicMethod);
+
+    jobject Q = (*env)->CallObjectMethod(env, publicKey, g_ECPublicKeyGetW);
+
+    (*env)->DeleteLocalRef(env, publicKey);
+
+    jobject xBn = (*env)->CallObjectMethod(env, Q, g_ECPointGetAffineX);
+    jobject yBn = (*env)->CallObjectMethod(env, Q, g_ECPointGetAffineY);
+
+    (*env)->DeleteLocalRef(env, Q);
+
+    // Success; assign variables
+    *qx = ToGRef(env, xBn);
+    *cbQx = CryptoNative_GetBigNumBytes(*qx);
+    xBn = NULL;
+    *qy = ToGRef(env, yBn);
+    *cbQy = CryptoNative_GetBigNumBytes(*qy);
+    yBn = NULL;
+    if (*cbQx == FAIL || *cbQy == FAIL)
+        goto error;
+
+    if (includePrivate)
+    {
+        jobject privateKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPrivateMethod);
+
+        if (!privateKey)
+        {
+            *d = NULL;
+            *cbD = 0;
+            goto error;
+        }
+
+        jobject dBn = (*env)->CallObjectMethod(env, privateKey, g_ECPrivateKeyGetS);
+
+        (*env)->DeleteLocalRef(env, privateKey);
+
+        *d = ToGRef(env, dBn);
+        *cbD = CryptoNative_GetBigNumBytes(*d);
+        if (*cbD == FAIL)
+            goto error;
+    }
+    else
+    {
+        if (d)
+            *d = NULL;
+
+        if (cbD)
+            *cbD = 0;
+    }
+
+    return SUCCESS;
+
+error:
+    *cbQx = *cbQy = 0;
+    *qx = *qy = 0;
+    if (d)
+        *d = NULL;
+    if (cbD)
+        *cbD = 0;
+    ReleaseLRef(env, xBn);
+    ReleaseLRef(env, yBn);
```

Comment:
Might be worth using @elinor-fung's VA_ARG pattern from the X509 version to make LRef cleanup more consistent.

Reply:
I'm personally not a fan of that pattern and prefer the explicit calls here. Especially since the ECC code has to use different cleanup methods depending on where the object was created from.

Reply:
Not going to push this but I would also recommend a similar mechanism. When there are this many locals that need management it is much harder to update and make sure one isn't missed.

Reply:
I'll give it another look then and see if I can make it work for the ECC code.